### ISSUE TITLE
Trouble reproducing reported training FPS

### ISSUE THREAD

**Issue by mx781** [google-research/seed_rl#80]
Hi there @lespeholt and maintainers,

I've been excited to try and make use of SEED for training on some custom environments, and although it is pretty fast, I seemed to be getting significantly lower FPS than reported in the paper. To get to the bottom of this, I tried reproducing the numbers of envs published in the paper. 

In particular, I've been running R2D2 on Atari (Pong) on a machine closest in spec to the V100 one listed in Table 1 in the paper. Instead of a V100, I have a RTX3090, and instead of 55 CPUs I have 64. With this setup I was expecting to match or exceed the 67K FPS listed in the table.

To stay faithful to the original code release, I ran everything off the original Docker image, except for a few tweaks to accommodate the dated image, namely fixing these req versions: `ale-py==0.7rc0 gym==0.18.3 opencv-python==3.4.14.53` (afaict these are the ones that would've been most recent at time of release). Matching up the hyperparams to Appendix A.3.1 (unspecified ones were already defaulting correctly), I have this command:
```bash
/run_local.sh atari r2d2 64 4 --total_environment_frames=2000000 --replay_ratio=0.75 --unroll_length=120 --replay_buffer_size=100000 --replay_buffer_min_size=5000 --inference_batch_size=64 --learning_rate=0.0001 --clip_norm=80
```

The actor-reported FPS in the logs seems to vary significantly, so instead I looked at various E2E measures (start/end of process tree, first/last actor log). The 20M run terminates in 19min59s, which comes out as **16,680 FPS -- roughly 4x slower than reported.** I later found that there's a `num_environment_frames/sec` logged to tensorboard, which hovered around 1.65e4, which aligns with my own measurements.

I've tried looking into a few possible culprits:
 - The original tf2.4.1 image was throwing "Value 'sm_86' is not defined for option 'gpu-name'", since TF2.4.x didn't support 8.6 compute capability. I bumped the image (and grpc) to work with 2.5.0 - that got rid of the error, but the FPS was the same.
 - I've confirmed TF is using the GPU, loading cuDNN, and using XLA and that I'm not somehow running in eager mode.
 - I've looked through the tensorboard metrics. They are fairly stable and seem reasonable, though I haven't found a reference point to compare the first two to.
     - elapsed_env_reset_s is around 3.3e-5 secs
     - elapsed_env_step_s is around 0.0016 secs
     - elapsed_inference_s is around 0.013 (this is 2x higher than reported in Table 13, but I suspect that number might be from the TPU runs?)
 - I've tried profiling/tracing the process (both in eager/non-eager mode) to see any obvious system level / io hiccups that might be causing this - nothing has come up.

In sum, I'm out of hints to investigate. Any advice? Happy to provide any further details needed.

Thanks in advance!

**Comment by lespeholt** [comment]
There could be a number of issues, differences in hardware/network and the code was generally written with TPUs in mind so maybe there has been some regression for GPUs in the code or in TensorFlow.

You could try and generate a trace which I always find is the easiest way to debug these issues. You can also try and basically replace the environment in the learner code with just zeros to verify that training is fast enough.  Finally you can try the IMPALA code which is somewhat simpler.
https://github.com/deepmind/scalable_agent
